helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java

```diff
@@ -250,6 +251,13 @@ public void dropInstance(String clusterName, InstanceConfig instanceConfig) {
     }
   }
 
+  @Override
+  public void purgeOfflineInstances(String clusterName, Long timeout) {
+    Map<String, InstanceConfig> timeoutOfflineInstances = findTimeoutOfflineInstances(clusterName
+        , timeout);
+    timeoutOfflineInstances.values().forEach(instance -> dropInstance(clusterName, instance));
```

Comment:
two high-level comments:
1. are we not using the lock? as discussed offline, i think to complete this feature, it would be a good idea to use the distributed lock here.
2. what kind of failure handling are we doing in this call? we're making multiple dropInstance calls here and there's no guarantee that all would go through. plus, the method itself isn't returning anything so there's no way for the user to determine whether the purge operation was successful or not

Reply:
1. if this is going to be called by the user, i think it's fair to assume that there would not be race conditions (because we expect the user to do concurrency control). however, i'm still a little confused.

> The major concern from user side of having locks in purge function and in instance joining function is that they may have thousands of instances come online at the same time, if each of them tries to read and write the lock, zookeeper will have a burst, and this is dangerous.

what we ultimately want to achieve is to provide a function to do this automatically. for now, we shift the burden of calling this to the user, but we want to automate this eventually. i'm not quite getting the "dangerous" part - why is it dangerous to use zk to implement the lock? it's one of the things zk is meant for. having hundreds and thousands of zk clients with zk as the distributed lock is not unheard of? i'm not exactly understanding the situation you're describing.

2. i was initially thinking maybe it might be better to catch the exception and throw it from this function as a "purge failure", with the list of instances that were purged successfully and (optionally) the list of instances that were not due to a failure if there are any. Otherwise, there's no way to know for the user or ourselves what exactly happened. I don't think that would be too hard to implement on top of what you already have now - do you think that would help us debug this better if something goes wrong?

Reply:
For question 1, Helix does not plan to integrate this purge logic into controller as it unnecessarily complicates the controller logic. Then it will be up to users to decide whether they would like take the risk for race condition, or afford the traffic in zookeeper. The concern from users was during deployment, zk is already undergoing a lot of traffic, and adding one more read/write per instance in a short time period (e.g. 10 or 20 seconds) for lock makes it worse without adding extra benefit for them. 
I think a possible future work could be Helix provides config to use lock or not during purge and instances come online, and users could choose to turn it on or off depending on their real use case.

Reply:
@zhangmeng916 let's read that one a little more carefully. 

> as discussed, i think what we ultimately want to achieve is to provide a service or a sidecar controller to do this automatically

this doesn't mean that we will include the purge logic into the main controller. it clearly states that it will be a service or a sidecar controller. and i believe we all agreed that this would be the direction.

the reason we are shifting the responsibility to the user is for the ease of implementation. ultimately, we want to automate this in our holistic cluster management environment (HaaS).

> adding one more read/write per instance in a short time period (e.g. 10 or 20 seconds) for lock makes it worse without adding extra benefit for them.

I'm not sure if i completely agree with the statement. 10-20 sec is a very generous duration where zk server will be handle many read/writes. also I don't agree that there's no extra benefit. the benefit we're talking about here is concurrency control in a distributed setting? i'm not suggesting we add read/write operations for nothing. 

if you prefer to leave out locking for this API (which is totally fine with the right assumptions), i suggest we give some context in the JavaDoc and clearly state that this operation is not thread-safe and that it _should_ lock in the future when called automatically.